src/assets.cc

```diff
@@ -0,0 +1,159 @@
+
+#include <blt/assets.hh>
+
+#if defined(BLT_USING_LIBCXX)
+
+#include <dlfcn.h>
+
+#include <dsl/FileSystem.hh>
+#include <dsl/Archive.hh>
+#include <dsl/DB.hh>
+#include <dsl/Transport.hh>
+
+#define hook_remove(hookName) SubHook::ScopedRemove _sh_remove_raii(&hookName)
+
+using namespace std;
+
+namespace blt {
+
+    typedef pair<dsl::idstring, dsl::idstring> hash_t;
+    std::map<std::pair<dsl::idstring, dsl::idstring>, std::string> custom_assets;
+
+    // LAPI stuff
+
+    namespace lapi { namespace assets {
+        int create_entry(lua_state *L) {
+            uint64_t extension = *(uint64_t*) lua_touserdata(L, 2);
+            uint64_t name = *(uint64_t*) lua_touserdata(L, 3);
+            hash_t hash(name, extension);
+
+            size_t len;
+            const char *filename_c = luaL_checklstring(L, 4, &len);
+            string filename(filename_c, len);
+
+            if(custom_assets.count(hash)) {
+                // Appears this is actually allowed in the basegame
+                // char buff[1024];
+                // snprintf(buff, 1024, "File already exists in replacement DB! %lx.%lx (%s)", name, extension, filename.c_str());
+                // luaL_error(L, buff);
+            }
+
+            custom_assets[hash] = filename;
+            return 0;
+        }
+
+        void setup(lua_state *L)
+        {
+#define func(name) \
+                lua_pushcclosure(L, name, 0); \
+                lua_setfield(L, -2, #name);
+
+            func(create_entry);
+
+#undef func
+        }
+    }; };
+
+
+    // The acual function hooks
+
+    SubHook dslDbAddMembers; // We need this to add the entries, as we need to do it after the DB table is set up
+
+    SubHook dslDbTryOpenDetour;
+    SubHook dslDbTryOpenDetour2;
+
+    void (*dsl_db_add_members)   (lua_state*);
+
+    void* (*dsl_db_try_open)   (void* /* this */, void* name, void* ext, void* target, void* transport, void* z);
+    void* (*dsl_db_try_open_2)   (void* /* this */, void* name, void* ext, void* target, void* transport, void* z);
+
+    int (*dsl_db_do_resolve_1) (dsl::DB *_this, dsl::idstring*, dsl::idstring*, void *abc, void *def);
+    void (*dsl_fss_open) (dsl::Archive *output, dsl::FileSystemStack **_this, std::string const*);
+
+    static void*
+    dt_dsl_db_try_open(void* _this, void* name, void* ext, void* target, void* transport, void* z) {
+        hook_remove(dslDbTryOpenDetour);
+
+        //printf("Hello, World: %ld %lx.%lx %ld\n", *((uint64_t*) name), *((uint64_t*) target), *((uint64_t*) ext), sizeof(std::string));
```

Comment:
Can you delete this debug output?

Reply:
Whoops, missed that.

Reply:
Doesn't appear in HEAD, GH says it's an outdated version.